File: packages/core/flutter/lib/src/debounce.dart
import 'dart:async';

/// A debounced function wrapper with a [cancel] method.
///
/// Invoke it like the original function; call [cancel] to discard any pending
/// execution without invoking the action (e.g., in [State.dispose]).
///
/// ```dart
/// final search = debounce<String>((query) => print('Searching: $query'));
/// search('hello');   // schedules execution after 250ms
/// search.cancel();   // discards the pending call
/// ```
class Debounced<T> {
  final void Function(T) _action;
  final Duration _wait;
  Timer? _timer;

  Debounced._(this._action, this._wait);

  /// Schedules [action] to be called with [arg] after the wait period.
  /// Resets the timer if called again before the wait elapses.
  void call(T arg) {
    _timer?.cancel();
    _timer = Timer(_wait, () => _action(arg));
  }

  /// Cancels any pending invocation without executing it.
  void cancel() {
    _timer?.cancel();
    _timer = null;
  }
}

/// Debounces a function call by delaying its execution until after a specified wait period.
/// Subsequent calls within the wait period reset the timer.
///
/// Returns a [Debounced] wrapper that can be called like the original function
/// and supports [Debounced.cancel] for cleanup.
///
/// **Parameters:**
/// - [action] - The function to debounce
/// - [wait] - Duration to wait before invoking the function (defaults to 250ms)
///
/// **Example:**
/// ```dart
/// final search = debounce<String>((String query) {
///   print('Searching: $query');
/// }, const Duration(milliseconds: 300));
///
/// search('hello'); // Will execute after 300ms if no other calls are made
/// search('world'); // Previous call is cancelled, this will execute after 300ms
/// search.cancel(); // Discards the pending call (e.g., in dispose())
/// ```
Debounced<T> debounce<T>(
  void Function(T) action, [
  Duration wait = const Duration(milliseconds: 250),
]) {
  return Debounced._(action, wait);
}

File: packages/core/web/src/debounce.ts
/**
 * A debounced function with a cancel method.
 * Call it like the original function; call `.cancel()` to discard any pending execution.
 */
export interface Debounced<T extends (...args: any[]) => void> {
  (...args: Parameters<T>): void;
  /** Cancels any pending invocation without executing it. */
  cancel(): void;
}

/**
 * Debounces a function call by delaying its execution until after a specified wait period.
 * Subsequent calls within the wait period reset the timer.
 * Returns a `Debounced` wrapper with a `cancel()` method for cleanup.
 *
 * @param fn - The function to debounce.
 * @param wait - Milliseconds to wait before invoking the function. Defaults to 250ms.
 * @returns A debounced wrapper with a `cancel()` method.
 *
 * @example
 * ```ts
 * const search = debounce((query: string) => {
 *   console.log('Searching:', query);
 * }, 300);
 *
 * search('hello'); // Will execute after 300ms if no other calls are made
 * search.cancel(); // Discards the pending call (e.g., on component unmount)
 * ```
 */
export function debounce<T extends (...args: any[]) => void>(
  fn: T,
  wait = 250,
): Debounced<T> {
  let timer: ReturnType<typeof setTimeout> | null = null;

  const debounced = (...args: Parameters<T>): void => {
    if (timer) clearTimeout(timer);
    timer = setTimeout(() => fn(...args), wait);
  };

  debounced.cancel = (): void => {
    if (timer) {
      clearTimeout(timer);
      timer = null;
    }
  };

  return debounced as Debounced<T>;
}
